libs/ballot-interpreter/src/hmpb-rust/timing_marks/corners/corner_finding.rs

```diff
@@ -0,0 +1,272 @@
+use image::RgbImage;
+use imageproc::drawing::draw_filled_rect_mut;
+use itertools::Itertools;
+use types_rs::geometry::{Point, Size};
+
+use crate::{
+    ballot_card::Geometry,
+    debug::imageproc_rect_from_rect,
+    image_utils::rainbow,
+    impl_cornerwise,
+    interpret::Error,
+    scoring::UnitIntervalScore,
+    timing_marks::{
+        corners::{
+            util::{mark_distances_to_point, CornerWise, EdgeWise},
+            BallotGridCandidateMarks,
+        },
+        CandidateTimingMark, Corner,
+    },
+};
+
+/// Represents the four corners of a ballot grid.
+#[derive(Debug, Clone)]
+pub struct BallotGridCorners {
+    top_left: BallotGridCorner,
+    top_right: BallotGridCorner,
+    bottom_left: BallotGridCorner,
+    bottom_right: BallotGridCorner,
+}
+
+impl_cornerwise!(BallotGridCorners, BallotGridCorner);
+
+const MIN_CORNER_TIMING_MARK_SCORE: UnitIntervalScore = UnitIntervalScore(0.8);
+
+impl BallotGridCorners {
+    /// Get the four corner candidate timing marks for the ballot grid.
+    pub fn corner_marks(
+        &self,
+    ) -> (
+        &CandidateTimingMark,
+        &CandidateTimingMark,
+        &CandidateTimingMark,
+        &CandidateTimingMark,
+    ) {
+        (
+            &self.top_left.best_corner_grouping.corner_mark,
+            &self.top_right.best_corner_grouping.corner_mark,
+            &self.bottom_left.best_corner_grouping.corner_mark,
+            &self.bottom_right.best_corner_grouping.corner_mark,
+        )
+    }
+
+    /// Get the top-left corner of the ballot grid.
+    pub const fn top_left(&self) -> &BallotGridCorner {
+        &self.top_left
+    }
+
+    /// Get the top-right corner of the ballot grid.
+    pub const fn top_right(&self) -> &BallotGridCorner {
+        &self.top_right
+    }
+
+    /// Get the bottom-left corner of the ballot grid.
+    pub const fn bottom_left(&self) -> &BallotGridCorner {
+        &self.bottom_left
+    }
+
+    /// Get the bottom-right corner of the ballot grid.
+    pub const fn bottom_right(&self) -> &BallotGridCorner {
+        &self.bottom_right
+    }
+
+    /// Find all corners of the ballot grid. Searches based on the left and
+    /// right edges rather than the top and bottom ones.
+    pub fn find_all(
+        image_size: Size<u32>,
+        geometry: &Geometry,
+        candidates: &BallotGridCandidateMarks,
+    ) -> Result<Self, Error> {
+        let timing_mark_center_to_center_distance =
+            geometry.timing_mark_size.width + geometry.timing_mark_size.height;
```

Comment:
This is a little confusing. It's not immediately obvious why the timing marks would be vertically spaced based on width.

Also, IIRC the distance between timing marks is currently not consistent across ballot image lengths, since we use flex containers to spread them out and the number of marks we're using doesn't divide evenly. Not sure if that will pose a larger issue to this algorithm. I see there's an error tolerance, but I figured it was worth calling out.

Reply:
That's a great point. I'll address this by computing the expected distance between timing marks in both dimensions.